### ISSUE TITLE
Swallowing of errors

### ISSUE THREAD

**Issue by lostpebble** [reframejs/wildcard-api#3]
Hi there,

Loving the simplicity of this library. I've just got a couple issues I'd like some clarification on.

How do I catch errors that occur inside my async function for my endpoint.

For example:

```typescript
import { endpoints } from "wildcard-api";

endpoints.runThing = async function({ argument }) {
  doSomethingWithArgument(argument); <-- This throws an Error
};
```

These errors are being swallowed and handled by Wildcard API, and the only result I get back from my request is:

```
HTTP/1.1 400 Bad Request
Content-Type: text/plain; charset=utf-8
Content-Length: 51
Date: Tue, 05 Mar 2019 14:29:25 GMT
Connection: keep-alive

{"usageError":"Endpoint could not handle request."}

Response code: 400 (Bad Request); Time: 49ms; Content length: 51 bytes
```

I would like to wrap my entire API in my own error handling, as I return certain structures of JSON from all my API endpoints for my client to handle as it likes. This includes uncaught errors in my API code - as of now I have no way to intercept that error and return my own structured response.

**Comment by lostpebble** [comment]
To add to this, I created a function which wraps each API function to catch any errors and return my structured response, but this is not the best because of a couple things:

* I need to remember to wrap every single method
* I can't seem to find a way to change the response status code

On the second point there, how would we usually do that? Is there a way to set a custom status code on a wildcard endpoint response? I've tried `this.statusCode = 401` and `this.status = 401`, but they don't seem to affect the response.

**Comment by brillout** [comment]
Any error thrown on the server shouldn't be propagated to the browser as this would be a security issue.

While not shown in the browser, errors are still shown in Node.js and outputted to your Node.js process' stderr. Basically: errors are swallowed from the perspective of the browser but aren't swallowed from the perspective of Node.js.

What's the reason you want to have access to errors in the browser? Are these validation errors?

Wildcard's use case is custom APIs. While error status codes make sense for a public API such as GitHub or Facebook's public API, error status code make less sense for custom APIs.

The idea here is that you pass information over the returned value of your API functions instead of fiddling around with HTTP status code.

**Comment by brillout** [comment]
Btw thanks for opening the tickets! :-) I'm glad that you like the simplicity of Wildcard and I hope you still like it after my reply :-)

**Comment by lostpebble** [comment]
@brillout the reason I want to catch the errors is that I catch all errors for all my APIs, and I return a specific structure to my client, like so:

```
{
  positive: boolean;
  endMessage: string;
  endTags: string[];
  payload: any;
}
```

I agree with you about not propagating server error messages to the client. In this case I would just return a very generic message like `An error occurred`.

But there is now no way for me to structure unknown thrown error responses to my client how I like while using Wildcard API - just as I've wrapped all other endpoints and thrown errors on my server with this structure to achieve consistency.

It would be nice if I could set an option on Wildcard when initialising it to tell it that I would like to handle any thrown errors completely.

And yea, no problem! I'm happy to contribute how I can 🙂 This kind of simplified, single-endpoint API was something I've been thinking about in my mind for some months now as well after finding GraphQL quite cool in its single endpoint concept, but just way too verbose and controlling at times. Your implementation seems really nice from what I've played with so far!

**Comment by brillout** [comment]
Ok makes sense, I'll look into this.

**Comment by brillout** [comment]
Would the following work out for you?

```js
const wildcardApi = require('wildcard-api');

widcardApi.onNewEndpointResult = ({endpointName, endpointArgs, endpointResult}) => {
  if( endpointResult instanceOf Error ) {
     this.statusCode = 401;
     this.body = "<html>some html</html>";
  } else {
     return endpointResult;
  }
};
```

**Comment by lostpebble** [comment]
Yea I guess that's one way of doing it - would definitely be able to intercept and do things there. Would I be getting the original error instance there? Would `endpointResult` be the actual thrown error inside my endpoint?

The reason I ask is because I have some internal tasks that might purposefully throw errors with certain structures (as a way of deep escape hatching) - and I would then like to be able to get the information directly off them like so (just an example, but would be something like this):

```typescript
if( endpointResult instanceOf TaskFunctionError) {
     this.statusCode = TaskFunctionUtil.getStatusCodeForError(endpointResult.errorPayload.endId);
     this.type = "application/json";
     this.body = TaskFunctionUtil.errorPayload;
  }
```

**Comment by brillout** [comment]
Yes that's the plan.

Also, `getApiResponse` will return the same object than the `this` object in your endpoint functions and in `onNewEndpointResult`. So you'll be able to access `this.type` in your Koa integration. E.g.:

~~~js
WildcardApiRouter.all("/wildcard/*", async ctx => {
  const { body, statusCode, type } = await getApiResponse(ctx);
  ctx.body = body;
  ctx.status = statusCode;
  ctx.type = type; // Assuming Koa understands ctx.type (I didn't look into Koa's documentation)
});
~~~

**Comment by lostpebble** [comment]
Okay, sounds good. I think that'll be a pretty great solution!

**Comment by brillout** [comment]
Yes I like it as well, thanks a lot for the brainstorming!

**Comment by brillout** [comment]
Done and published as `wildcard-api@0.1.2` (https://www.npmjs.com/package/wildcard-api).

Check out https://github.com/brillout/wildcard-api/blob/master/test/tests/onNewEndpointResult.js to see how to use `onNewEndpointResult`. See the args `endpointResut` and `endpointError`. Also add `ctx.type = apiResponse.type;` to your koa integration.

Let me know if this works for you.

Keeping this ticket open as I'll write proper docs for the readme later.

**Comment by lostpebble** [comment]
Great! Thanks, will be checking this out today and report back.

**Comment by brillout** [comment]
Have you encountered a problem with Wildcard? Let me know if there is anything that led you away from Wildcard. I'm curious:).

**Comment by lostpebble** [comment]
Hey @brillout . Nope not at all - I've just moved on to other parts of my stack for the time being. I built out a small app when I made this issue and haven't had the opportunity to try out the new error capabilities - but will soon.

Something extra that would be cool that I've been thinking of lately would be some kind of wildcard web socket interface for live client data. But that can be a separate issue for a later time :)

**Comment by brillout** [comment]
Ok:)

> Something extra that would be cool that I've been thinking of lately would be some kind of wildcard web socket interface for live client data. But that can be a separate issue for a later time :)

Yes, that would be super cool.

If you want we can design and implement this together... We could create a new GitHub organisation and we would both be owner of it. Let me know if that's something you'd be up to. Maybe we can make Wildcard to an exciting adventure :-).

Let me know and then I'll make the code more readable. (I've not focused on code readability while I implemented the first Wildcard prototype.)

**Comment by lostpebble** [comment]
I'd be up for that! I do currently have lots of development baggage on my current projects though (I run a small startup as a single developer - and I'm very stretched as it is), so I don't know how consistently I'll be able to help out. But that said, I'm open to give it a go! I do really like what you've done so far, and think it could be something great.

I don't think there is actually much missing besides the websockets layer... but of course, there are always rooms for improvement.

I just wish more people would try it out. Marketing is always the tough thing with projects, especially one like this where people are already quite stuck in their ways with server -> client communication.

**Comment by brillout** [comment]
I guess you should focus on your startup then:)

Ideas how a real time interface would look like? If you're busy then simply don't answer my question ;).

How about simply inverting the direction of function definition <-> function call:

~~~js
// Node.js
// Instead of defining functions on the server, we can also define them in the browser
// Essentially making Wildcard bi-directional, aka real-time
const ret = await endpoints.iAmDefinedInTheBrowser();
assert(ret===42);
~~~

~~~js
// browser
endpoints.iAmDefinedInTheBrowser = function() {
  alert('I have been called "directly" from the browser.');
  return 42;
};
~~~

Btw about your startup; Wildcard can be a selling point to attract devS. E.g. my work on Reframe has made it easier for me to hire devS.

And yea the main to-do for Wildcard is to somehow getting people to know about Wildcard. Not sure how to approach this yet. I guess I could apply for talks.

**Comment by lostpebble** [comment]
Interesting... I like the direction you are going there. So basically, if I call that function on Node.js the browser will execute the function on its side? (and I can pass variables in to that function too, I guess?).

The only issue I see with this which just came to me - and I think its a big one - is that a server -> client communication is one -> many. Real-time communication needs to be able to open a channel of communication specific to a single client who is listening for specific data. How would that work in this example?

Its a tough one. I think you might have to wrap something around a websocket server. This one looks very good: https://github.com/uNetworking/uWebSockets.js .

> Btw about your startup; Wildcard can be a selling point to attract devS. E.g. my work on Reframe has made it easier for me to hire devS.

Unfortunately we're running very lean at the moment. But yea, when the time comes I will definitely show off all the cool projects I've been making use of! ;) I'm sure it'll help attract devs.

**Comment by brillout** [comment]
> Interesting... I like the direction you are going there. So basically, if I call that function on Node.js the browser will execute the function on its side? (and I can pass variables in to that function too, I guess?).

Yes exactly. So same thing but the other way around.

>  server -> client communication is one -> many

Was thinking about something like:

~~~js
// Node.js
const {endpoints} = require('wildcard-api');

const roomClients = {};

endpoints.joinChatRoom = function(roomId) {
  roomClients[roomId] = roomClients[roomId] || [];
  roomClients[roomId].push(this.clientId);
};

endpoints.createNewMessage = function(roomId, msgText) {
  const clientsInRoom = roomClients[roomId];

  if( !clientsInRoom.includes(this.clientId) ) {
     throw new Error(
       'Permission denied: '+
       'you cannot post a message to a chat room you did not join.'
     );
  }

  // `clients` is a context option to set which clients will execute the function
  const context = {clients: clientsInRoom};

  endpoints.onNewMessage.bind(context)(msgText);
};

~~~

~~~js
// Browser
const {endpoints} = require('wildcard-api/client');

endpoints.onNewMessage = function(msgText) {
  console.log('New message: ' + msgText);
};

const chatRoomId = 12345;
await endpoints.joinChatRoom(chatRoomId);
await endpoints.createNewMessage(chatRoomId, 'hello to everyone');
~~~

Didn't know about uWebSockets, seems nice, thanks!

**Comment by brillout** [comment]
> Interesting... I like the direction you are going there. So basically, if I call that function on Node.js the browser will execute the function on its side? (and I can pass variables in to that function too, I guess?).

Yes exactly. So same thing but the other way around.

>  server -> client communication is one -> many

Was thinking about something like:

~~~js
// Node.js
const {endpoints} = require('wildcard-api');

const roomClients = {};

endpoints.joinChatRoom = function(roomId) {
  roomClients[roomId] = roomClients[roomId] || [];
  roomClients[roomId].push(this.clientId);
};

endpoints.createNewMessage = function(roomId, msgText) {
  const clientsInRoom = roomClients[roomId];

  if( !clientsInRoom.includes(this.clientId) ) {
     throw new Error(
       'Permission denied: '+
       'you cannot post a message to a chat room you did not join.'
     );
  }

  // `clients` is a context option to set which clients will execute the function
  const context = {clients: clientsInRoom};

  endpoints.onNewMessage.bind(context)(msgText);
};

~~~

~~~js
// Browser
const {endpoints} = require('wildcard-api/client');

endpoints.onNewMessage = function(msgText) {
  console.log('New message: ' + msgText);
};

const chatRoomId = 12345;
await endpoints.joinChatRoom(chatRoomId);
await endpoints.createNewMessage(chatRoomId, 'hello to everyone');
~~~

Didn't know about uWebSockets, seems nice, thanks!

**Comment by lostpebble** [comment]
I think that looks pretty cool. The only thing that might be confusing is having both of those things on the `endpoints` namespace.

As it is now, we have been using `endpoints` on the client to call code on the server. Now you're using the same thing to call code on the client. I think it would be better to "namespace" the methods that call back to the client under `clientEndpoints` or something:

```ts
// on server
clientEndpoints.onNewMessage.bind(context)(msgText);

// on client
clientEndpoints.onNewMessage = function(msgText) {
  console.log('New message: ' + msgText);
};
```

**Comment by lostpebble** [comment]
Unrelated to the current discussion (but actually on topic to the issue lol), I've just updated wildcard API to the latest version but I can't seem to get the error interception to work.

My code looks like this, but that `console.log("Intercepting endpoint result");` is never being called:

```ts
import * as wildcard from "wildcard-api";
import { TaskFunctionError, TaskFunctionUtils } from "@gt/task-function-utils/build/TaskFunctionUtils";
import { TFRFailure, TFRFailureDefaults, TFRSuccess } from "@gt/task-function-utils/build/TaskFunctionResponses";
import { ETaskFunctionEndId } from "@gt/task-function-utils/build/TaskFunctionTypes";
import Router from "koa-router";

export const WildcardApiRouter = new Router();

WildcardApiRouter.all("/wildcard/*", async ctx => {
  const { body, statusCode, type } = await wildcard.getApiResponse(ctx);
  ctx.type = type;
  ctx.body = body;
  ctx.status = statusCode;
});

wildcard.endpoints.fail = function() {
  throw new Error(`Something horrible happened here...`);
};

wildcard.endpoints.failTaskFunction = function() {
  throw new TaskFunctionError(TFRFailure(ETaskFunctionEndId.UNAUTHORIZED, `Testing unauthorized error throw`));
};

wildcard.endpoints.okay = function() {
  return TFRSuccess();
};

wildcard.onNewEndpointResult = function({endpointName, endpointArgs, endpointResult, endpointError}) {
  console.log(`Intercepting endpoint result`);

  if (endpointError) {
    this.type = "application/json";

    if (endpointError.constructor === TaskFunctionError) {
      console.log(`Got task function error`);
      this.statusCode = TaskFunctionUtils.getHttpStatusCodeForEndId((endpointError as TaskFunctionError).taskFunctionResponse.endId);
      this.body = (endpointError as TaskFunctionError).taskFunctionResponse;
    } else {
      this.statusCode = 401;
      this.body = TFRFailureDefaults({ endMessage: `Something went wrong on Wildcard endpoint "${endpointName}": ${endpointError.message}`, errorPayload: endpointError });
    }
  }
};
```

As far as I can tell the code is exactly like you have it in your tests... So I'm a little confused as to what I'm doing wrong here. Any ideas?

**Comment by brillout** [comment]
- Make sure to use the latest version `wldcard-api@0.1.3`
 - I did a repro of your code and it works: https://github.com/brillout/repro/tree/wildcard-lostpebble
 - Feel free to create a repro and I'll take a look
 - Feel free to share me your code and I'll take a look
 - I'll reply to the rest later

**Comment by brillout** [comment]
@lostpebble The API of `onNewEndpointResult` has changed in `wildcard-api@0.1.4`. You can use `overwriteResponse` and `overwriteResult` instead, see https://github.com/brillout/wildcard-api/blob/master/test/tests/onNewEndpointResult.js.

**Comment by brillout** [comment]
@lostpebble I renamed `onNewEndpointResult` to `onEndpointCall` in `wildcard@0.2.0`. It's now documented at https://github.com/brillout/wildcard-api#onEndpointCall.

**Comment by brillout** [comment]
I'm closing this, feel free to re-open if something doesn't work.

**Comment by brillout** [comment]
In my efforts to simplify Wildcard I've redesigned how Wildcard handles errors. The new version `wildcard-api@0.4.0` still includes `onEndpointCall` but I'd like to deprecate it.

I believe there should be a way to achieve what you want without `onEndpointCall`.

> But there is now no way for me to structure unknown thrown error responses to my client how I like while using Wildcard API

Maybe you could wrap all your endpoint functions:

~~~js
const {endpoints} = require('wildcard-api');

endpoints.myFirstEndpoint = catchUnexpectedErrors(async function(name) {
  return {greeting: "Hello "+name};
});

endpoints.mySecondEndpoint = catchUnexpectedErrors(async function() {
  const data = await db.getData();
  return data;
});

function catchUnexpectedErrrors(endpointFunction) {
  return function(...args) {
       try {
         return endpointFunction(...args);
       } catch(err) {
          return {
             errorMessage: "There is a bug",
          };
       }
   }
}
~~~

And on the browser side you can then do whatever you want:

~~~js
const {endpoints} = require('wildcard-api/client');

(async () => {
  const ret = await endpoints.mySecondEndpoint();
  if( ret.errorMessage ) {
     // Do what your non-wildcard endpoints would do here
  }
})();
~~~

If I'm not mistaken, this should give you all the flexibility you'd need.

Let me know if that works for you.

Infos about the new error handling design: https://github.com/reframejs/wildcard-api#error-handling

**Comment by lostpebble** [comment]
Hi @brillout 

Okay cool. I'll stick with the wrapping of my endpoints then, as I mentioned earlier.

It's not the greatest thing though to be honest, as people will expect their endpoints to throw errors completely instead of them being swallowed by wildcard - as one would expect if we were creating regular API endpoints. Is there a reason why you need to take control of catching the errors?

As I mentioned earlier regarding wrapping:

> To add to this, I created a function which wraps each API function to catch any errors and return my structured response, but this is not the best because of a couple things:
>
> * I need to remember to wrap every single method
> * I can't seem to find a way to change the response status code

How would I deal with status code things now?

I think what you state in your error docs is not exactly true:

> On the server, your endpoint functions should not throw errors. And if one of your endpoint function does throw an error, then it should be a bug.

I have plenty of backend code which throws errors on purpose given certain situations - which are obviously caught directly in my routing library and the appropriate status codes etc. are returned. I mentioned my scenario eariler about `TaskFunctionError` (https://github.com/reframejs/wildcard-api/issues/3#issuecomment-471273069). Throwing errors can be a great way to "short-circuit" deep code without having to create failure checks all the way back up again.

---

That said, its not the end of the world if I can still control my response (status codes etc.) using the wrapping method. How would this be achieved?

**Comment by brillout** [comment]
> people will expect their endpoints to throw errors completely instead of them being swallowed by wildcard

Agreed and Wildcard actually never swallows errors. All uncaught errors happening in your endpoint function are always printed in Node.js's stderr.

I find libraries that swallow errors super annoying. Swalling errors is a no-go.

That said, and as we agreed earlier, errors should not propagate to the browser.

> I have plenty of backend code which throws errors on purpose given certain situations

Sorry, I wasn't very clear: I was speaking of uncaught errors. Caught errors are perfectly fine.


> Is there a reason why you need to take control of catching the errors?

Yes, to make Wildcard easier. My first and foremost goal with Wildcard is to make it easy. I'm fairly happy with the ease of https://github.com/reframejs/wildcard-api#error-handling (Note that re-rewrote the whole section yesterday. It should be easier to read now.) But maybe I'm missing something here.


> How would I deal with status code things now?

Don't deal with them at all? I believe that Wildcard abstracting away the HTTP protocol from you is a good thing. I can't image a situation where you'd need to deal with status code. By wrapping your endpoint functions and wrapping your endpoint calls, I believe you can achieve anything you want without ever dealing with HTTP stuff.

> I need to remember to wrap every single method

The trade off here is simplicity VS convenience. Agreed, wrapping every endpoint function is annoying. But on the other hand adding not-strictly-needed features to Wildcard makes Wildcard less simple. Not sure what is more important here but right now I'm leaning more toward simplicity.

Instead of wrapping every endpoint function, maybe you can wrap the function calls that throw the expected errors you want to catch in the first place?

**Comment by lostpebble** [comment]
Okay cool, I will continue using wildcard with the wrapping of my endpoints and catch as needed.

Will let you know if any use-cases arise where I run into some friction.

**Comment by lostpebble** [comment]
Hi Romuald

Running into this kind of cryptic error today.

```
****************************************
************* Stack Trace **************
****************************************
    at getCallStack (C:\dev\_projects\vibescout-web-workspace\node_modules\@wildcard-api\server\node_modules\@brillout\assert\assert.js:169:27)
    at Function.assert (C:\dev\_projects\vibescout-web-workspace\node_modules\@wildcard-api\server\node_modules\@brillout\assert\assert.js:24:21)
    at Function.module.exports [as internal] (C:\dev\_projects\vibescout-web-workspace\node_modules\@wildcard-api\server\node_modules\@brillout\assert\internal.js:9:19)
    at Object.get (C:\dev\_projects\vibescout-web-workspace\node_modules\@wildcard-api\server\WildcardApi.js:355:14)
    at Proxy.<anonymous> (C:\dev\_projects\vibescout-web-workspace\workspaces\vs-cloud-functions\functions\grow_traction_shipping_dash\dist\index.js:1250:27)
    at runEndpoint (C:\dev\_projects\vibescout-web-workspace\node_modules\@wildcard-api\server\WildcardApi.js:144:39)
    at getApiHttpResponse (C:\dev\_projects\vibescout-web-workspace\node_modules\@wildcard-api\server\WildcardApi.js:63:53)
    at GTSDash_CloudFunction_httpFunction (C:\dev\_projects\vibescout-web-workspace\workspaces\vs-cloud-functions\functions\grow_traction_shipping_dash\dist\index.js:1474:84)
    at C:\dev\_projects\vibescout-web-workspace\workspaces\vs-cloud-functions\node_modules\@google-cloud\functions-framework\build\src\invoker.js:100:17


****************************************
************ Internal Error ************
****************************************
Failed assertion condition: `false != true`

Error thrown by endpoint function `getUser`.
Error is printed above.
Read the "Error Handling" documentation for how to handle errors.
```

I understand your logic behind handling and returning from errors to keep things simple. But please can you do a `console.error(error)` with the actual error object so we can get proper server logs with the stack and everything included? Its MUCH easier to dig through the code when the native error log takes care of the errors instead of having to deal with a static string like you output here.

You can still continue do the extra stuff, such as displaying the message you have here, but I think its really important that at least one native `console.error(err)` happens when the core error occurs so that we can quickly deal with it in a way that our tooling etc. is more able to interpret.

**Comment by brillout** [comment]
I'm looking at it now

**Comment by brillout** [comment]
Wildcard does(/should) always show the original error. And actually the error you are seeing *is* the original error. The error comes from the Wildcard source code, not from your code. Seems to be a bug. Let me see if I can get more infos from the stack trace alone.

**Comment by lostpebble** [comment]
So I managed to fix the error itself - its more that the error message was a bit cryptic and didn't help much, especially cause the stack isn't easy to find what's wrong.

I was trying to access `this.session` without actually having set up my session stuff on this project yet.

**Comment by brillout** [comment]
>  the error message was a bit cryptic and didn't help much

I'm sorry about that and it's a bug on my part. I do care about nice error messages.

I'm writing a fix & unit test.

I'm not forgetting the donation-reminder thingy, I've just postponed writing a fix it but I do care that you can upgrade to the latest Wildcard version sooner rather than later.

**Comment by lostpebble** [comment]
> Wildcard does(/should) always show the original error. And actually the error you are seeing _is_ the original error. The error comes from the Wildcard source code, not from your code. Seems to be a bug. Let me see if I can get more infos from the stack trace alone.

I get that, and for the most part its been fine, but it would be great if you could just `console.error(error)` that original error as well instead of just statically outputting a formatted error like you do at the moment. This allows my software to easily take me to the correct place in my code files, with source maps etc. as well just by clicking on the parts of the stack output - because that is the native way that they expect the errors to be presented.

**Comment by lostpebble** [comment]
Another important thing that I just realized that could cause overload of information - is that you seem to be logging multiple `console.error()` logs, as seen here in the code:

```js
function logError({ err, endpointName }) {
  console.error("");
  console.error(err);
  console.error("");
  console.error(
    colorizeError(
      "Error thrown by endpoint function" +
        (!endpointName ? "" : " `" + endpointName + "`") +
        "."
    )
  ) + console.error("Error is printed above.");
  console.error(
    'Read the "Error Handling" documentation for how to handle errors.'
  );
  console.error("");
}
```

This could have the side-effect of causing my error tracking libraries to think each of those individual lines are something that needs to be logged and looked at. I think it could be better to just log a single error of the original error object, and then maybe use `console.warn()` or something else for the further information.

**Comment by brillout** [comment]
That's actually the case. Wildcard doesn't format errors originating from your code.

~~~js
    try {
      endpointResult = await endpoint.apply(requestProps_proxy, endpointArgs);
    } catch (err) {
      endpointError = err;
    }
~~~
Then Wildcard `console.error` the original error `endpointError` without any modifications.

Wildcard only changes the format of its own errors. Is that a problem for you? Your software doesn't help much for errors originating from Wildcard source code I guess?

**Comment by brillout** [comment]
About `console.error`: it's important that `Read the "Error Handling" documentation for how to handle errors.` is shown to `stderr` and not `stdout` (the user should not be able to miss that message).

I can replace all `console.error('')` with `console.log('')` and make sure there are only two `console.error` calls: one for the original error, and a second one for the message telling the user he should look at the docs. Would that be acceptable to you that your software will believe there are 2 errors although there is only one?

**Comment by brillout** [comment]
Actually you are right `console.warn` is better than `console.error` for further information. I'll implement this.

**Comment by lostpebble** [comment]
> Then Wildcard console.error the original error endpointError without any modifications.

Oh, okay. I must have been mistaken then, just thought that I'd seen this formatted type error before like this and figured all of them were like that when something goes wrong inside an endpoint.

It's not a big problem if Wildcard changes the format of its own internal errors. Its a little unexpected to get a wall of text with that kind of formatting though.

> I can replace all console.error('') with console.log('') and make sure there are only two console.error calls: one for the original error, and a second one for the message telling the user he should look at the docs. Would that be acceptable to you that your software will believe there are 2 errors although there is only one?

Yea, that's fine. Why are you using the console for making spaces though? You could just use a template string with all the correct spacing set inside it? or just some `/n`.

The bottom line is that I just want the key information about what's wrong - a simple error log of the actual error so I can dig into the line numbers easily and quickly understand the error messages and such as its presented in a way that is expected. I also would prefer if "log spam" was kept to a minimum as much as possible, so preferably not so many seemingly unnecessary log calls only to create spacing. And added formatting should be kept to a minimum, as characters = data and sometimes we have millions of logs going into a backend log storage etc. where superfluous info can add up quickly and make it hard to get to the "meat" of the logged information.

Thanks for being so responsive btw!

**Comment by brillout** [comment]
There is a legacy reason why Wildcard formats its own errors but I don't think it's required anymore, I'll reevaluate this. And you're right I should and I will reduce chars and `console` calls to a minimum. Thanks!

**Comment by brillout** [comment]
I'm in the middling of changing the prod architecture of the Lsos which has high prio. I will implement the Wildcard stuff after I'm done, ETA being this WE. (I'm quite excited about the new prod architecture, I'll tell you about it on our next Paris walk ;-).)

**Comment by brillout** [comment]
tmrw

**Comment by lostpebble** [comment]
Okay cool. Interested to hear what you have put together on the next walk :)

**Comment by brillout** [comment]
Fixed in `@wildcard-api/server@0.3.2`. I added a test to ensure that passing `undefined` results in an empty context object instead of throwing.

Note that you can upgrade to `@wildcard-api/server@0.3.2` while pinning `@wildcard-api/client` to a version before the donation-reminder.

**Comment by brillout** [comment]
@lostpebble 

I just published `@wildcard-api/client@0.5.3` and `@wildcard-api/server@0.5.3` which:
 - Includes the improved error messages I mentioned previously.
 - Removes the donation-reminder.

It does however include the `lsos` package (https://github.com/Lsos/lsos-js) which as well does have a postinstall script but it timeouts command calls and will `process.exit()` and let you know what command didn't terminate.

I couldn't reproduce the GCE bug using the docker image.

> Seems like it contradicts the MIT license, so not sure how that works legally.

The MIT License explicitly allows selling Software:
~~~
Permission is hereby granted [...] the rights to [...] sell copies of the Software.
~~~

> how that might end up with some unwanted "friendly fire" to smaller devs

No activation key is required if:
 - The user's repository has only few active Git authors.
 - The user's repository is public.

This means the only friendly fire is that some companies will have to waste 5 minutes to get a free activation key. I expect such false positives to be fairly rare; if a repository is developed by many authors, it's like within a larger company.

Lsos projects can also opt-in to never block their users but to show a `console.warn` instead. This means that users can indefinitely use the Lsos project without activation key. This makes the whole thing a trust-based thing.

> I also don't too much enjoy the idea of a whole bunch of extra code 

Me too. The `lsos` package is small and only calls Git commands to do all its checks. No network/external calls beyond Git commands.

**Comment by brillout** [comment]
@lostpebble 

I'm going to add a `setErrorHandler` interface.

~~~ts
import { setErrorHandler } from 'telefunc/server';

setErrorHandler(myErrorHandler);

// `myErrorHandler` will be called anytime one of your telefunction throws an error.
function(err) {
   // handle the error `err` thrown by one of your telefunction.
}
~~~

Allowing to do all kinds of things.

~~~js
import { setErrorHandler } from 'telefunc/server';

setErrorHandler(function(err) {
   // This is what Telefunc does by default
   console.error(err);
   // Telefunc automatically constructs a sensible HTML response (404, 400, or 500)
});
~~~
~~~js
import { setErrorHandler } from 'telefunc/server';

setErrorHandler(function(err) {
   // The Telefunc server middleware will re-throw the error.
   // This means that it is your server framework (Express/Koa/Hapi/...) that will handle the error
   throw err;
});
~~~
~~~js
import { setErrorHandler } from 'telefunc/server';
import { logError } from './my-error-logger';

setErrorHandler(function(err) {
   console.error(err);
   logError(err);
});
~~~

Let me me know if you object on anything.

Hope everything's good on your side.
